aws/src/test/java/org/apache/iceberg/aws/s3/S3InputStreamTest.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import com.adobe.testing.s3mock.junit4.S3MockRule;
+import com.amazonaws.services.s3.AmazonS3URI;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Random;
+import org.apache.commons.io.IOUtils;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertThrows;
+
+public class S3InputStreamTest {
+  @ClassRule
+  public static final S3MockRule S3_MOCK_RULE = S3MockRule.builder().silent().build();
+
+  private final S3Client s3 = S3_MOCK_RULE.createS3ClientV2();
+  private final Random random = new Random(1);
+
+  @Before
+  public void before() {
+    s3.createBucket(CreateBucketRequest.builder().bucket("bucket").build());
+  }
+
+  @Test
+  public void testRead() throws Exception {
+    AmazonS3URI uri = new AmazonS3URI("s3://bucket/path/to/read.dat");
+    byte[] expected = randomData(1024 * 1024);
+
+    writeS3Data(uri, expected);
+
+    try (InputStream in = new S3InputStream(s3, uri)) {
+      byte[] actual = IOUtils.readFully(in, expected.length);
+      assertArrayEquals(expected, actual);
+    }
```

Comment:
I think this needs to test the different read methods, not just read with a buffer. It also should test the values returned by `getPos` because those are critical to skipping correctly for read-through seeks.

Reply:
I added a number of additional tests including backseeks, read-through, and skip seeks and reads using both implemented paths.  This was actually good because it caught an issue with our lazy position updates for the streams.